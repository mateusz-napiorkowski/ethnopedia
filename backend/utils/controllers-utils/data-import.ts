export interface subData {
    name: string
    values: string
    subcategories: Array<any>
}

export const prepRecords: any = (data: any, collectionName: any) => {
    const header = data[0]
    const recordsData = data.slice(1)
    const headerAttrsInArrays: any = []
    header.forEach((attr:string) => {
        headerAttrsInArrays.push(attr.split("."))
    })

    const records: Array<[]> = []
    for(const recordIndex in recordsData) {
        const recordAttrs = recordsData[recordIndex]
        const newRecord: any = {categories: []}
        const depth = 1
        const categories: any = newRecord.categories
        recordAttrs.forEach((cellVal: any, attrIndex: number) => {
            if(header[attrIndex].split(".").length === depth) {
                const fields: any = []
                header.forEach((element: any) => {
                    if(element.startsWith(header[attrIndex]) && element.split(".").length === depth + 1) {
                        fields.push(element)
                    }
                });
                const newCaterory: subData = {
                    name: header[attrIndex], 
                    values: recordAttrs[attrIndex].toString().split(";").filter((i: any) => i !== ""),
                    subcategories: []
                }
                if(fields.length !== 0) {
                    newCaterory.subcategories = fillSubcategories(depth + 1, fields, recordAttrs, header, recordsData, recordIndex)
                }
                categories.push(newCaterory)
            }
        });
        newRecord.collectionName = collectionName
        records.push(newRecord)
    }
    return records
}

export const fillSubcategories: any = (depth: number, fields: any, allAttrs: any, header: any, recordsData: any, recordIndex: number) => {
    const subs: any = []
    let deeperFields: any = []
    fields.forEach((field: any) => {
        deeperFields = []
        header.forEach((attrName: string) => {
            if(attrName.startsWith(field) && attrName.split(".").length === depth + 1) {
                deeperFields.push(attrName)
            }
        });
        const newSub: subData = {name: field.split(".").slice(-1)[0],
            values: recordsData[recordIndex][header.indexOf(field)].toString().split(";").filter((i: any) => i !== ""),
            subcategories: []
        }
        if(deeperFields.length !== 0) {
            newSub.subcategories = fillSubcategories(depth + 1, deeperFields, allAttrs, header, recordsData, recordIndex)
        }
        subs.push(newSub)
    });
    return subs
}