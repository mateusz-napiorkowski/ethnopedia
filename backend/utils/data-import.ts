import mongoose, { isValidObjectId } from "mongoose"
import { artworkCategory, record } from "./interfaces"
import path from "path"
import fs from "fs";
import unzipper from "unzipper"

const validateCategories = async (header: string[], asNewCollection: boolean) => {
    if(asNewCollection) {
        const categories = header.filter(categoryName => categoryName.trim() !== "_id" && categoryName.trim() !== "nazwy plików")
        const missingCategories = categories.filter((category: string) => !header.includes(category))
        const unnecessaryCategories = header.filter((category: string) => {
            if(category === "_id" || category === "nazwy plików")
                return false
            return !categories.includes(category)
        })
        if(missingCategories.length != 0 || unnecessaryCategories.length != 0) {
            throw new Error(`Brakujące kategorie: ${missingCategories}, Nadmiarowe kategorie: ${unnecessaryCategories}`)
        }
    }
    if(header.length !== new Set(header).size)
        throw new Error ("Header has duplicate values")
    if(header[header.length - 1] == '')
        throw new Error ("Row contains more columns than the header")
    if(header.includes(""))
        throw new Error (`Header has empty fields`)     
    for(const categoryName of header) {
        const categoryNameSplitByDot = categoryName.split('.')
        if(categoryNameSplitByDot.includes(""))
            throw new Error (`No subcategory name after the dot symbol in header field: ${categoryName}`)
        const directParentCategoryFullName = categoryNameSplitByDot.slice(0,-1).join('.')
        if(categoryNameSplitByDot.length > 1 && !header.includes(directParentCategoryFullName))
            throw new Error (`Missing parent category: '${directParentCategoryFullName}'`)
    }
}

const prepUploadsDirAndArchiveBuffer = async (zipFile: Express.Multer.File | undefined, collectionId: string) => {
    if(!zipFile) return {}

    const uploadsDir = path.join(__dirname, "..", `uploads/`);
    if (!fs.existsSync(uploadsDir)) fs.mkdirSync(uploadsDir);
    const collectionUploadsDir = path.join(__dirname, "..", `uploads/${collectionId}`);
    if (!fs.existsSync(collectionUploadsDir)) fs.mkdirSync(collectionUploadsDir);

    const archiveBuffer = await unzipper.Open.buffer(zipFile.buffer);

    return {archiveBuffer, collectionUploadsDir}
}

const processArchiveFiles = (
    newRecord: any, archiveBuffer: any, oldRecordId: string, newRecordId: string,
    filenamesCell: string | undefined, collectionUploadsDir: any,
    collectionId: string, zipFile: Express.Multer.File | undefined
) => {
    let uploadedFilenames = []
    let failedUploadsCauses = []
    if(filenamesCell && zipFile) {
        newRecord.files = []
        const maxFileSize = 25 * 1024 * 1024 // 25 MB
        const archiveFilesData = filenamesCell.trim()
            .split(";")
            .filter(Boolean)
            .map(item => {
                const [index, filename] = item.split(":");
                const ext = path.extname(filename);
                const newFilename = `${newRecordId.toString()}_${index}${ext}`
                return {
                    oldFileName: `${oldRecordId}_${index}${ext}`,
                    newFilename: newFilename,
                    userFilename: filename
                };
            });
        for(const entry of archiveFilesData) {
            const fileEntry = archiveBuffer!.files.find((d: any) => d.path === entry.oldFileName);
            if(fileEntry) {
                try {
                    if(!/\.(mei|mid|midi|txt|text|musicxml|mxl|xml|wav|mp3)$/i.test(entry.oldFileName))
                        throw Error("Invalid file extension")
                    if(fileEntry.uncompressedSize > maxFileSize)
                        throw Error("File size exceeded")
                    const outputPath = path.join(collectionUploadsDir!, entry.newFilename);
                    const writeStream = fs.createWriteStream(outputPath);
                    fileEntry.stream().pipe(writeStream);
                    const fileProps = {
                        originalFilename: entry.userFilename,
                        newFilename: entry.newFilename,
                        filePath: `uploads/${collectionId}/${entry.newFilename}`,
                        size: fileEntry?.uncompressedSize,
                        uploadedAt: Date.now()
                    }
                    newRecord.files.push(fileProps)
                    uploadedFilenames.push(entry.oldFileName)
                } catch (error) {
                    const err = error as Error
                    failedUploadsCauses.push({archiveFilename: entry.oldFileName, userFilename: entry.userFilename, cause: err.message})
                }
            } else {
                failedUploadsCauses.push({archiveFilename: entry.oldFileName, userFilename: entry.userFilename, cause: "File not found in the archive"})         
            }
            
        }
    }
    return {uploadedFilenames, uploadedFilesCount: uploadedFilenames.length, failedUploadsCauses}
}

const setRecordCategories = (row: string[], newRecord: any, header: string[]) => {
    row.forEach((categoryValueUntrimmed, columnIndex) => {
        if(header[columnIndex] !== "_id" && header[columnIndex] !== "nazwy plików") {
            const categoryValue = categoryValueUntrimmed.trim()
            const isNotSubcategoryColumn = header[columnIndex].split(".").length === 1
            if(isNotSubcategoryColumn) {
                const directSubcategoriesNames = header.filter(columnName => 
                    columnName.startsWith(`${header[columnIndex]}.`) && columnName.split(".").length === 2
                )
                newRecord.categories.push({
                    name: header[columnIndex],
                    value: categoryValue,
                    subcategories: fillSubcategories(directSubcategoriesNames, 2, header, row)
                })      
            }
        }  
    });
}

export const prepRecordsAndFiles = async (
    data: Array<Array<string>>, collectionName: string,
    asNewCollection: boolean, collectionId: string, zipFile?: Express.Multer.File | undefined
) => {
    try {
        const header = data[0]
            .map(categoryName => categoryName.trim().replace(/\s*\.\s*/g, '.'))
        validateCategories(header, asNewCollection)

        const {collectionUploadsDir, archiveBuffer} = await prepUploadsDirAndArchiveBuffer(zipFile, collectionId)

        const recordsData = data.slice(1)
        const records: Array<record> = []
        const _idColumnIndex = header.indexOf("_id")
        const filenamesColumnIndex = header.indexOf("nazwy plików")
        let totalUploadedFilesCount = 0
        let totalFailedUploadsCauses = []
        let allUploadedFilenames = []
        for(const row of recordsData) {
            const oldRecordId = isValidObjectId(row[_idColumnIndex].trim()) ? row[_idColumnIndex].trim() : undefined;
            const newRecordId = !oldRecordId || asNewCollection 
                ? new mongoose.Types.ObjectId()
                : new mongoose.Types.ObjectId(oldRecordId)
            const newRecord: record = {_id: newRecordId, categories: [], collectionName: collectionName, files: []}

            const {uploadedFilesCount, uploadedFilenames, failedUploadsCauses} = processArchiveFiles(
                newRecord,
                archiveBuffer,
                oldRecordId!,
                newRecordId.toString(),
                filenamesColumnIndex ? row[filenamesColumnIndex] : undefined,
                collectionUploadsDir,
                collectionId,
                zipFile
            )

            setRecordCategories(row, newRecord, header)
            
            records.push(newRecord)
            allUploadedFilenames.push(...uploadedFilenames)
            totalUploadedFilesCount += uploadedFilesCount
            totalFailedUploadsCauses.push(...failedUploadsCauses)
        }
        const allArchiveFilenames = archiveBuffer ? archiveBuffer.files.map(file => file.path) : [];
        const unlistedFiles = allArchiveFilenames.filter(x => !allUploadedFilenames.includes(x))
        for(const file of unlistedFiles) {
            totalFailedUploadsCauses.push({archiveFilename: file, cause: "File is not associated with any record"})
        }
        return {
            records,
            uploadedFilesCount: totalUploadedFilesCount,
            failedUploadsCount: totalFailedUploadsCauses.length,
            failedUploadsCauses: totalFailedUploadsCauses
        }
    } catch (error) {
        throw new Error("Invalid data in the spreadsheet file", {cause: error})
    }
    
}

const fillSubcategories = (fields: Array<string>, depth: number, header: Array<string>, rowValues: Array<string>) => {
    const subcategories: Array<artworkCategory> = []
    fields.forEach(field => {
        const allSubcategoriesNames = header.filter(columnName => (columnName.startsWith(`${field}.`)))
        const directSubcategoriesNames = allSubcategoriesNames.filter(columnName => columnName.split(".").length === depth + 1)
        subcategories.push({
            name: field.split(".").slice(-1)[0],
            value: rowValues[header.indexOf(field)].trim(),
            subcategories: fillSubcategories(directSubcategoriesNames, depth + 1, header, rowValues)
        })        
    });
    return subcategories
}