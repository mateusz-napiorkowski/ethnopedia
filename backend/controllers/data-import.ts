import { Request, Response } from "express"
import Artwork from "../models/artwork";
import { authAsyncWrapper } from "../middleware/auth"
import { getNewRecordIdsMap, handleFilesUnzipAndUpload, prepRecords } from "../utils/data-import";
import CollectionCollection from "../models/collection";
import mongoose, { ClientSession } from "mongoose";
import { findMissingParentCategories, transformCategoriesArrayToCategoriesObject } from "../utils/categories";

export const importData = authAsyncWrapper(async (req: Request, res: Response) => {
    try {
        if(!req.body.importData || req.body.importData.length < 2 || !req.body.collectionId )
            throw new Error("Incorrect request body provided")
        const session = await mongoose.startSession()
        await session.withTransaction(async (session: ClientSession) => {
            const collectionId = req.body.collectionId
            const foundCollections = await CollectionCollection.find({_id: collectionId}, null, {session}).exec()
            if (foundCollections.length !== 1 )
                throw new Error(`Collection not found`)
            const collectionName = foundCollections[0].name!
            const records = await prepRecords(req.body.importData, collectionName, false, collectionId, undefined)
            const bulkWriteOps = records.map(record => ({
                updateOne: {
                    filter: { _id: record._id, },
                    update: { $set: record },
                    upsert: true
                }
            }));
            const result = await Artwork.bulkWrite(bulkWriteOps, {session});
            return res.status(201).json(result)
        });
        session.endSession()
    } catch (error) {
        const err = error as Error
        console.error(error)
        if (err.message === `Incorrect request body provided`)
            res.status(400).json({ error: err.message })
        else if (err.message === `Collection not found`)
            res.status(404).json({ error: err.message })
        else if (err.message === "Invalid data in the spreadsheet file"){
            res.status(400).json({ error: err.message, cause: err.cause?.toString()})}
        else
            res.status(503).json({ error: `Database unavailable` })
    }
})
 
export const importDataAsCollection = authAsyncWrapper(async (req: Request, res: Response) => {
    try {
        let importData;        
        try {
            importData = JSON.parse(req.body.importData);
        } catch {
            throw new Error(`Incorrect request body provided`);
        }
        const collectionName = req.body.collectionName
        const description = req.body.description
        const zipFile = req.file
        if(!importData || importData.length < 2 || !collectionName || !description )
            throw new Error("Incorrect request body provided")
        if(zipFile && !/\.(zip)$/i.test(zipFile.originalname))
            throw Error("Invalid file extension")
        const session = await mongoose.startSession()
        await session.withTransaction(async (session: ClientSession) => {
            const categoriesArray = importData[0].filter((cat: string) => cat !== "_id").map((category: string) => category.trim())
            const missingCategories = findMissingParentCategories(categoriesArray)
            if(missingCategories.length !== 0)
                throw new Error(
                    "Invalid categories data",
                    {cause: `Brakujące kategorie nadrzędne: ${missingCategories.toString()}`}
                )
            const categories = transformCategoriesArrayToCategoriesObject(categoriesArray)
            const newCollection = await CollectionCollection.create([
                {name: collectionName, description: description, categories: categories}
            ], {session})

            const newRecordIdsMap = getNewRecordIdsMap(importData)
            const records = await prepRecords(importData, collectionName, true, newCollection[0]._id.toString(), newRecordIdsMap, zipFile)
            const result = await Artwork.insertMany(records, {session})
            return res.status(201).json({newCollection, result})
        });
        session.endSession()
    } catch (error) {
        const err = error as Error
        console.error(error)
        if (err.message === `Incorrect request body provided` || err.message === `Invalid file extension`)
            res.status(400).json({ error: err.message })
        else if (err.message === "Invalid data in the spreadsheet file" || err.message === "Invalid categories data"){
            res.status(400).json({ error: err.message, cause: err.cause })}
        else
            res.status(503).json({ error: `Database unavailable` })
    }
})