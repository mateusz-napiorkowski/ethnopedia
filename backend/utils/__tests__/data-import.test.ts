import {describe, expect, test, jest, beforeEach} from "@jest/globals"
import {prepRecords} from "../data-import"

const mockGetAllCategories = jest.fn()
jest.mock("../categories", () => ({
    getAllCategories: () => mockGetAllCategories(),
}))

describe('data-import controller util functions tests', () => {
    beforeEach(() => {
        jest.resetAllMocks()
    })

    test.each([
        {
            testName: 'prepRecords test - three levels deep data',
            data: [
                ['Title', 'Title.Subtitle', 'Title.Subtitle.Subsubtitle', 'Artists'],
                ['title 1', 'subtitle 1',
                'subsubtitle 1'],
            ],
            getAllCategories: ['Title', 'Title.Subtitle', 'Title.Subtitle.Subsubtitle', 'Artists']
        },
        {
            testName: 'prepRecords test - unusual column order',
            data: [ ['Title.Subtitle', 'Title', 'Title.Subtitle.Subsubtitle'],
            ['subtitle 1', 'title 1', 
                'subsubtitle 1']],
            getAllCategories: ['Title.Subtitle', 'Title', 'Title.Subtitle.Subsubtitle']
        },
        {
            testName: 'prepRecords test - skipped category values in some records',
            data: [
                ['Title', 'Artists', 'Release Year'],
                ['title 1', '', 'release year 1'],
                ['', 'artists 2', 'release year 2'],
                ['title 3', '', ''],
            ],
            getAllCategories: ['Title', 'Artists', 'Release Year']
        },
        {
            testName: 'prepRecords test - trim all whitespace category values to empty string',
            data: [
                ['Title', 'Title.Subtitle', 'Artists'],
                ['title 1', ' ', ' ',]

            ],
            getAllCategories: ['Title', 'Title.Subtitle', 'Artists']
        },
        {
            testName: 'prepRecords test - remove leading/trailing whitespace in each part of full category name',
            data: [
                [' Title', 'Title . Subtitle ', '  Artists   '],
                ['title 1', 'subtitle 1', 'artists 1',]

            ],
            getAllCategories: ['Title', 'Title.Subtitle', 'Artists']
        },
        {
            testName: 'prepRecords test - remove leading/trailing whitespace in category values',
            data: [
                ['Title', 'Title.Subtitle', 'Artists'],
                ['  title 1   ', '  subtitle 1   ', '    artists 1    ',]

            ],
            getAllCategories: ['Title', 'Title.Subtitle', 'Artists']
        },
    ])(`$testName`,
        async ({data, getAllCategories}) => {
            mockGetAllCategories.mockReturnValue(getAllCategories)
            expect(await prepRecords(data,
                "collection", false
            )).toMatchSnapshot()
        }
    )

    test.each([
        {
            testName: 'prepRecords test - throw error when missing categories in header when importing to already existing collection',
            data: [
                ['Title', 'Artists'],
                ['title 1', 'artist 1'],
            ],
            getAllCategories: ['Title', 'Artists', 'Year'],
            asCollection: false
        },
        {
            testName: 'prepRecords test - throw error when unnecessary categories in header when importing to already existing collection',
            data: [
                ['Title', 'Artists', "Year"],
                ['title 1', 'artist 1', "1999"],
            ],
            getAllCategories: ['Title', 'Artists'],
            asCollection: false
        },
        {
            testName: 'prepRecords test - throw error when duplicate names in header',
            data: [
                ['Title', 'Title.Subtitle', 'Artists', 'Title.Subtitle'],
                ['title 1', 'subtitle 1', 'artists 1', 'subtitle 1'],
            ],
            getAllCategories: undefined,
            asCollection: true
        },
        {
            testName: 'prepRecords test - throw error when row contains more columns than the header',
            data: [
                ['Title', 'Artists', ''],
                ['title 1', 'artists 1', 'error causing value'],
            ],
            getAllCategories: undefined,
            asCollection: true
        },
        {
            testName: 'prepRecords test - throw error when header has empty fields',
            data: [
                ['Title', '', 'Artists'],
                ['title 1', 'some value', 'artists 1'],
            ],
            getAllCategories: undefined,
            asCollection: true
        },
        {
            testName: 'prepRecords test - throw error when parent category of some subcategory is missing',
            data: [
                ['Title', 'Title.MissingSubcategory.Subtitle'],
                ['title 1', 'subtitle 1'],
            ],
            getAllCategories: undefined,
            asCollection: true
        },
        {
            testName: 'prepRecords test - throw error when no subcategory name after the dot symbol in header field',
            data: [
                ['Title', 'Title.'],
                ['title 1', 'some value'],
            ],
            getAllCategories: undefined,
            asCollection: true
        },
        {
            testName: 'prepRecords test - throw error when subcategory name is only whitespace',
            data: [
                ['Title', 'Title. '],
                ['title 1', 'some value'],
            ],
            getAllCategories: undefined,
            asCollection: true
        }
    ])(`$testName`,
        ({data, getAllCategories, asCollection}) => {
            mockGetAllCategories.mockReturnValue(getAllCategories)
            expect(async () => await prepRecords(data,
                "collection", asCollection
            )).rejects.toThrow("Invalid data in the spreadsheet file")
        }
    )
})