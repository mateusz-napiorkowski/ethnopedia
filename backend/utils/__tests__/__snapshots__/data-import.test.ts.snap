// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`data-import controller util functions tests prepRecords test - remove leading/trailing whitespace in category values 1`] = `
[
  {
    "categories": [
      {
        "name": "Title",
        "subcategories": [
          {
            "name": "Subtitle",
            "subcategories": [],
            "values": [
              "subtitle 1",
              "alternate subtitle 1",
            ],
          },
        ],
        "values": [
          "title 1",
          "alternate title 1",
        ],
      },
      {
        "name": "Artists",
        "subcategories": [],
        "values": [
          "artists 1",
        ],
      },
    ],
    "collectionName": "collection",
  },
]
`;

exports[`data-import controller util functions tests prepRecords test - remove leading/trailing whitespace in each part of full category name 1`] = `
[
  {
    "categories": [
      {
        "name": "Title",
        "subcategories": [
          {
            "name": "Subtitle",
            "subcategories": [],
            "values": [
              "subtitle 1",
            ],
          },
        ],
        "values": [
          "title 1",
        ],
      },
      {
        "name": "Artists",
        "subcategories": [],
        "values": [
          "artists 1",
        ],
      },
    ],
    "collectionName": "collection",
  },
]
`;

exports[`data-import controller util functions tests prepRecords test - skip all whitespace category values 1`] = `
[
  {
    "categories": [
      {
        "name": "Title",
        "subcategories": [],
        "values": [
          "title 1",
        ],
      },
    ],
    "collectionName": "collection",
  },
]
`;

exports[`data-import controller util functions tests prepRecords test - skip subcategory with empty values only if there are no deeper subcategories 1`] = `
[
  {
    "categories": [
      {
        "name": "Title",
        "subcategories": [
          {
            "name": "Subtitle",
            "subcategories": [
              {
                "name": "Subsubtitle",
                "subcategories": [
                  {
                    "name": "Subsubsubtitle",
                    "subcategories": [],
                    "values": [
                      "subsubsubtitle 1",
                    ],
                  },
                ],
                "values": [],
              },
            ],
            "values": [
              "subtitle 1",
            ],
          },
        ],
        "values": [],
      },
    ],
    "collectionName": "collection",
  },
  {
    "categories": [
      {
        "name": "Title",
        "subcategories": [
          {
            "name": "Subtitle",
            "subcategories": [
              {
                "name": "Subsubtitle",
                "subcategories": [
                  {
                    "name": "Subsubsubtitle",
                    "subcategories": [
                      {
                        "name": "Subsubsubsubtitle",
                        "subcategories": [],
                        "values": [
                          "subsubsubsubtitle 1",
                        ],
                      },
                    ],
                    "values": [],
                  },
                ],
                "values": [],
              },
            ],
            "values": [],
          },
        ],
        "values": [],
      },
    ],
    "collectionName": "collection",
  },
]
`;

exports[`data-import controller util functions tests prepRecords test - skipped category values in some records 1`] = `
[
  {
    "categories": [
      {
        "name": "Title",
        "subcategories": [],
        "values": [
          "title 1",
        ],
      },
      {
        "name": "Release Year",
        "subcategories": [],
        "values": [
          "release year 1",
        ],
      },
    ],
    "collectionName": "collection",
  },
  {
    "categories": [
      {
        "name": "Artists",
        "subcategories": [],
        "values": [
          "artists 2",
        ],
      },
      {
        "name": "Release Year",
        "subcategories": [],
        "values": [
          "release year 2",
        ],
      },
    ],
    "collectionName": "collection",
  },
  {
    "categories": [
      {
        "name": "Title",
        "subcategories": [],
        "values": [
          "title 3",
        ],
      },
    ],
    "collectionName": "collection",
  },
]
`;

exports[`data-import controller util functions tests prepRecords test - three levels deep data 1`] = `
[
  {
    "categories": [
      {
        "name": "Title",
        "subcategories": [
          {
            "name": "Subtitle",
            "subcategories": [
              {
                "name": "Subsubtitle",
                "subcategories": [],
                "values": [
                  "subsubtitle 1",
                  "alternate subsubtitle 1",
                  "another alternate subsubtitle 1",
                ],
              },
            ],
            "values": [
              "subtitle 1",
              "alternate subtitle 1",
              "another alternate subtitle 1",
            ],
          },
        ],
        "values": [
          "title 1",
          "alternate title 1",
        ],
      },
    ],
    "collectionName": "collection",
  },
]
`;

exports[`data-import controller util functions tests prepRecords test - unusual column order 1`] = `
[
  {
    "categories": [
      {
        "name": "Title",
        "subcategories": [
          {
            "name": "Subtitle",
            "subcategories": [
              {
                "name": "Subsubtitle",
                "subcategories": [],
                "values": [
                  "subsubtitle 1",
                  "alternate subsubtitle 1",
                  "another alternate subsubtitle 1",
                ],
              },
            ],
            "values": [
              "subtitle 1",
              "alternate subtitle 1",
              "another alternate subtitle 1",
            ],
          },
        ],
        "values": [
          "title 1",
          "alternate title 1",
        ],
      },
    ],
    "collectionName": "collection",
  },
]
`;
